XCTestのテスト対象を変更した

Hyperbolic4183 opened the pull request:
## 目的
XCTestの対象をViewからModel、Viewを含めたViewControllerに変更すること
## 実装の概要
- `Fetchable`プロトコルに準拠した`FetcherMock`を作成し、'WeatherViewController'に注入している。
- 'WeatherViewController'はViewの更新に非同期処理を含んでいるので、'XCTestExpectation'を使った。
- `addProcessing(target: Void,completion: @escaping () -> Void)`を定義し、ボタンを押した後に、テスト項目の処理を加えられるようにした。
## レビューして欲しいところ
- `addProcessing(target: Void,completion: @escaping () -> Void)`を定義する以外に非同期処理をテストするための実装が思い浮かばなかった。ボタンを押した後に非同期処理が入る場合、どのようにテストをするのが良いか教えて欲しい。

Hyperbolic4183 committed:
fix 修飾子を変更した

WeatherViewはテストで用いるためprivate(set) varに、WeatherModelはprivateをつけてアクセスできる範囲を狭めた。

Hyperbolic4183 committed:
update モックを作成

Fetchableに準拠したモデルを作成した。テスト専用のクラスなので、積極的にイニシャライザを作った。

Hyperbolic4183 committed:
update getUI部品関数を修正

取得に失敗した時にすぐにわかるようにオプショナルをなくした。引数の型を変更した。

Hyperbolic4183 committed:
update テストの対象にモデルを含めるようにした。

WeatherViewControllerに非同期処理が存在しているので、XCTestExpectationを使用した。時間は4秒と設定しているが、実際にはすぐに終わる。

Hyperbolic4183 committed:
update 弱参照に変更

ykws reviewed (commented):
まず Button タップを実装し直すのは、非同期処理が冗長なため、しなくて良いと思いました。
現状規定してもらっている protocol ベースでテストしたいシナリオを以下に整理してみます。

### テストしたいシナリオ
- Fetchable の fetch を通して天気予報の取得を要求する
- FetchableDelegate の fetch を経由して天気予報の結果を取得する
- WeatherView に天気予報の結果が表示される

`WeatherViewController` に対してのアクセスを制限しているのと、
`FetchableDelegate` で ViewController を Mock できるのでテストしたいシナリオを以下のように確認できると良さそうです。

- FetchableDelegate を実装して ViewController を Mock する
  - ViewController Mock に検証用の WeatherView を持たせる
  - FetchableDelegate の fetch を実装して WeatherView を更新する
  - ViewController Mock の WeatherView を検証する

ViewController を Mock するので非同期のことは意識しないで済むようになります。

Hyperbolic4183 committed:
update WeatherViewControllerのモックを作成

FetchableDelegateを準拠させたWeatherViewControllerMockを作成 refs #33

Hyperbolic4183 committed:
update 晴れ画像に関するテストを変更

ViewControllerをモック化しテストを行うようにした。refs #33

Hyperbolic4183 committed:
update 全てのテストをViewControllerをモック化して行うようにした。

ViewControllerをモック化してテストを行うようにした。refs #33

Hyperbolic4183 committed:
fix テストのために変更していた修飾子を戻した

refs #33

ykws commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -83,4 +96,62 @@ class Yumemi_ios_trainingTests: XCTestCase {
+    }
+    
+    func fetch(_ fetchable: Fetchable?, didFailWithError error: WeatherAppError) {
+        
```
![memo-badge](https://img.shields.io/badge/review-next-blueviolet)

エラーパターンもテストできると良いです

ykws replied:
## エラーのシナリオ
- Fetchable の fetch を通して天気予報の取得を要求する
- FetchableDelegate の fetch を経由して天気予報のエラーを取得する

WeatherView に対する更新は特にないため、
エラーを返す Fetchable の Mock を用意して、 FetchableDelegate でエラーの処理が呼ばれることを確認できれば良いと思います。呼ばれるかどうかは XCTestExpectation が使えます。

現状はエラーの表示が ViewController に依存しているので、
今の時点はこれくらいで良いと思いますが、
この部分も Controller から分離して View のテストができるようにすると良いです。

ykws commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -83,4 +96,62 @@ class Yumemi_ios_trainingTests: XCTestCase {
+    
+    let weather: WeatherInformation.Weather
+    let minTemperature: String
+    let maxTemperature: String
```
![imo-badge](https://img.shields.io/badge/review-imo-orange.svg)

`weatherInformation` 以外は参照しないので `init` で直接 `WeatherInformation` に渡してしまえば定義しなくても良いです

ykws commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -9,7 +9,7 @@ import UIKit
 class WeatherViewController: UIViewController {
     private let weatherView = WeatherView()
-    private var weatherModel: Fetchable
+    private(set) var weatherModel: Fetchable
```
![nits-badge](https://img.shields.io/badge/review-nits-green.svg)

ViewController Mock しているので変更不要になってますか？

ykws commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -10,60 +10,73 @@ import XCTest
-        XCTAssertEqual(weatherImage, UIImage(named: "sunny"))
+        
+        fetcherMock = FetcherMock(weather: .sunny)
+        weatherViewControllerMock = WeatherViewControllerMock(model: fetcherMock)
+        weatherViewControllerMock.model.fetch()
```
![imo-badge](https://img.shields.io/badge/review-imo-orange.svg)

Model と ViewController は依存していないので、無理に持たせなくても良いです
一例として、

```suggestion
        weatherViewControllerMock = WeatherViewControllerMock()
        fetcherMock.delegate = weatherViewControllerMock
        fetcherMock.fetch()
```

ykws reviewed (approved):
いくつかコメントはしましたが、 protocol ベースでテストできているので良いと思いました！

![LGTM](https://image.lgtmoon.dev/90452)

novr commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -10,60 +10,73 @@ import XCTest
-        let weatherImage = getImage(from: view)
-        XCTAssertEqual(weatherImage, UIImage(named: "sunny"))
+        
+        fetcherMock = FetcherMock(weather: .sunny)
+        weatherViewControllerMock = WeatherViewControllerMock(model: fetcherMock)
```
[ASK] テスト対象は何でしょうか？
FetcherMockの結果を使ってWeatherViewControllerMockをテストしています

Hyperbolic4183 committed:
clean モッククラスに専用のファイルを用意した

WeatherViewControllerMockとSuccessFetcherMock専用のファイルを用意した。FetcherMockはSuccessFetcherMockに名前を変更した。refs #33

Hyperbolic4183 committed:
update エラーが起きるモデルモックを作成

Fetchableに準拠したfetch()でエラーが起こるモックFailedFetcherMockを作成 refs #33

Hyperbolic4183 committed:
fix 修飾子を変更

モッククラスを用意したため、setを外した。 refs#33

Hyperbolic4183 committed:
update XCTExpectationを拡張

エラーが起きた際に処理する関数が実行されることを確認する時に用いるプロパティを定義

Hyperbolic4183 committed:
update テストの書き方を変更

refs #33

Hyperbolic4183 committed:
update エラーに関するテストを追加

XCTestExpectationを使い、Fetcherがエラーを起こした時にdelegateメソッドが実行される事を確認するテストを作成 refs #33

ykws commented on Yumemi-ios-trainingTests/WeatherViewControllerMock.swift:
```diff
@@ -0,0 +1,24 @@
+
+class WeatherViewControllerMock: FetchableDelegate {
+    
+    var model: Fetchable?
```
![nits-badge](https://img.shields.io/badge/review-nits-green.svg)

使ってないです

ykws commented on Yumemi-ios-trainingTests/FailedFetcherMock.swift:
```diff
@@ -0,0 +1,23 @@
+//  Created by 大塚 周 on 2021/09/03.
+//
+
+import Foundation
```
![nits-badge](https://img.shields.io/badge/review-nits-green.svg)

import しなくて大丈夫です

ykws commented on Yumemi-ios-trainingTests/SuccessFetcherMock.swift:
```diff
@@ -0,0 +1,24 @@
+//  Created by 大塚 周 on 2021/09/03.
+//
+
+import Foundation
```
![nits-badge](https://img.shields.io/badge/review-nits-green.svg)

import しなくて大丈夫です

ykws commented on Yumemi-ios-trainingTests/WeatherViewControllerMock.swift:
```diff
@@ -0,0 +1,24 @@
+//
+
+import XCTest
+import Foundation
```
![nits-badge](https://img.shields.io/badge/review-nits-green.svg)

import しなくて大丈夫です

ykws commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -83,4 +96,12 @@ class Yumemi_ios_trainingTests: XCTestCase {
         return minTemperatureLabel
     }
     
+    private func getReloadButton(from view: UIView) -> UIButton {
```
![nits-badge](https://img.shields.io/badge/review-nits-green.svg)

使わなくなってますかね？

ykws commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -83,4 +96,12 @@ class Yumemi_ios_trainingTests: XCTestCase {
+    
+}
+
+extension XCTestExpectation {
```
![imo-badge](https://img.shields.io/badge/review-imo-orange.svg)

なんだろうと思ってしまったので、
単に Delegate に対して expectation を渡して、 `fulfill()` すれば良いと思いました。

ykws commented on Yumemi-ios-trainingTests/SuccessFetcherMock.swift:
```diff
@@ -0,0 +1,24 @@
+        self.weatherInformation = weatherInformation
+    }
+    
+    func fetch() {
```
![imo-badge](https://img.shields.io/badge/review-imo-orange.svg)

成功と失敗の Mock を分けずに、 error のプロパティを持たせて、
エラーを注入できるようにし、エラーの場合に `didFailWithError` を呼んで良いと思います。

ykws commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -10,60 +10,73 @@ import XCTest
+        XCTAssertEqual(testMinTemperature, minTemperature.text)
+    }
+    
+    func test_Fetcherでエラーが起きた時にWeatherViewControllerのdelegateが動く() {
```
![suggestion-badge](https://img.shields.io/badge/review-suggestion-blue.svg)

この形式に合わせて表現を見直してみてください

`天気予報がsunnyだったときに画面に晴れ画像が表示される`

`天気予報が〇〇だったときに〇〇される`

Hyperbolic4183 committed:
clean Modelモックを統一

fetch失敗時のモックとfetch成功時のモックを一つのモックに統一した。#refs 33

Hyperbolic4183 committed:
update WeatherViewをI/Fに依存させた。

weatherViewStateにdidSetを設定し、ControllerからWeatherViewProtocolに準拠したViewのUI部品を変更できるようにした。 refs #33

Hyperbolic4183 committed:
NotificationCenterによってloadingViewの表示切り替えを行うようにした。

loadingViewの表示はプロトコルにつけるほどの機能ではないと判断し、WeatherView特有の機能とした。WeatherViewControllerのfetch開始/終了を通して通知を受け取る。 refs #33

Hyperbolic4183 committed:
update WeatherViewControllerをI/Fに依存させるようにした。

WeatherViewControllerProtocolを作成し、Controllerが委譲されているプロトコルを準拠させた。viewとmodelをプロパティとした。refs #33

Hyperbolic4183 committed:
update WeatherViewControllerのモックが準拠しているプロトコルを変更した

WeatherViewControllerMockについて、モデルであるFetcableから委譲されたプロトコルFetchableDelegateについての準拠するのではなく、FetchableDelegateを準拠したWeatherViewControllerProtocolに準拠するようにした。 refs #33

Hyperbolic4183 committed:
clean 不必要な関数と拡張を削除
refs #33

Hyperbolic4183 committed:
update ModelモックとControllerモックの変更に伴うテストの変更

refs #33

Hyperbolic4183 committed:
clean 修飾子と拡張プロトコルの明示

final修飾子をつけ、プロトコルの実装のための拡張には、そのプロトコルを書くようにした。 refs #33

Hyperbolic4183 committed:
update テスト名の変更

refs #33

Hyperbolic4183 committed:
update 関連性のないプロトコルの準拠をなくした。

refs #33

Hyperbolic4183 committed:
fix テストに必要なプロトコルのみ準拠させた

refs# 33

Hyperbolic4183 committed:
update エラーの処理をVCに依存させないようにした。

ErrorHandler構造体を作成し、テスト時にフェイクオブジェクトが注入できるようにした。refs #33

Hyperbolic4183 committed:
update キューの切り替えをDIできるようにした。

QueueScheduler構造体を作成し、イニシャライザに追加した。 refs #33

Hyperbolic4183 committed:
update didSetではなく関数を使って画面更新を行うようにした。

refs #33

Hyperbolic4183 committed:
update イニシャライザを変更した。

refs #33

Hyperbolic4183 committed:
update エラーのテストを追加した。

refs #33

Hyperbolic4183 committed:
キューの切り替えをプロトコルではなくDIによって行うようにしてテストを作成した。

refs #33

Hyperbolic4183 committed:
fix switchView()が画面更新時に行われないバグを修正

refs #33

Hyperbolic4183 committed:
clean 不必要な改行を削除

refs #33

novr commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -8,13 +8,28 @@
+        self.mainQueueScheduler = queueScheduler
+        self.errorHandler = errorHandler
+        super.init(nibName: nil, bundle: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(reload), name: UIApplication.didBecomeActiveNotification, object: nil)
```
[NITS] 今回の内容と別ですが
reloadメソッド内でViewへアクセスしているのでviewDidLoad以後に登録のがよいかと思います

novr commented on Yumemi-ios-trainingTests/FetcherMock.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    var weatherInformation: WeatherInformation!
+    var error: WeatherAppError!
+    let isFetchSucceed: Bool
```
[IMO] `Result<WeatherInformation, WeatherAppError>`型で表現できる

novr commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -8,13 +8,28 @@
         NotificationCenter.default.addObserver(self, selector: #selector(reload), name: UIApplication.didBecomeActiveNotification, object: nil)
     }
+    
+    init(model: Fetchable, queueScheduler: MainQueueScheduler, errorHandler: ErrorHandler) {
```
[IMO] default引数とすると1つに集約できそうです
```suggestion
    init(model: Fetchable, queueScheduler: MainQueueScheduler = .live, errorHandler: ErrorHandler = .presentAlertViewController) {
```

novr commented on Yumemi-ios-training/Controller/ErrorHandler.swift:
```diff
@@ -0,0 +1,34 @@
+
+fileprivate func extractedFunc(_ error: Error, _ viewController: UIViewController) {
+    let message: String
+    switch error as! WeatherAppError {
```
[MUST] WeatherAppErrorに強制キャストしている
WeatherAppError型に限定するか、unknownErrorにフォールバックすべき

novr commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -8,13 +8,28 @@
-    private var weatherModel: Fetchable
+
+    let weatherView = WeatherView()
+    var weatherModel: Fetchable
```
[IMO] Testで`weatherViewController.weatherModel.fetch()`するために公開範囲を変えるよりは
Testでは`fetcherMock.fetch()`を呼び出していいかと
`test_最低気温がUILabelに反映される()`ではfetcherMock.fetch()で対応している

novr commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -10,60 +10,81 @@ import XCTest
     }
     
-    private func getImage(from view: WeatherView) -> UIImage? {
+    private func getImage(from view: UIView) -> UIImage {
```
[MUST] XCTAssertEqualはnilとも比較してくれるので戻り値はUIImage?として、!しない

novr commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -10,60 +10,81 @@ import XCTest
     
-    private func getMaxTemperatureLabel(from view: WeatherView) -> UILabel {
+    private func getMaxTemperatureLabel(from view: UIView) -> UILabel {
         //WeatherView含まれる2つのUIStackViewのうち、2番目に追加したstackViewForLabelsを取得
```
[NITS] UI構造に依存してしまうのでaccessibilityIdentifierなど活用してもよいかと

novr commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -10,60 +10,81 @@ import XCTest
 
 class Yumemi_ios_trainingTests: XCTestCase {
     
+    var fetcherMock: FetcherMock!
```
[IMO] test間を超えてアクセスしたりsetup等でインスタンス化するわけではないので、メソッド内変数でよさそうです

Hyperbolic4183 committed:
update モデルにアクセス修飾子を追加。

refs #33

Hyperbolic4183 committed:
update 変数のアクセス範囲を狭くした。

refs #33

Hyperbolic4183 committed:
update 不必要な部分を削除

refs #33

Hyperbolic4183 committed:
fix 強制キャストを無くした。

refs #33

Hyperbolic4183 committed:
update Result型を使って可読性を高めた。

refs #33

Hyperbolic4183 committed:
update accessibilityIdentifier専用のstaticプロパティを作成した。

refs #33

Hyperbolic4183 committed:
update イニシャライザにデフォルト値を設定した

refs #33

Hyperbolic4183 committed:
update UI部品の取得方法を変更した

テストファイルでのみ対象のViewの拡張を行いUI部品を取得するようにした。 refs #33

Hyperbolic4183 committed:
update Controllerのプロパティにprivate修飾子を追加した。

全てprivateにし、堅牢性を高めた。refs #33

Hyperbolic4183 committed:
fix コメントを修正した

refs #33

novr commented on Yumemi-ios-trainingTests/Yumemi_ios_trainingTests.swift:
```diff
@@ -11,76 +11,128 @@ import XCTest
+    var stackViewForImageViewAndLabels: UIStackView? {
+        return self
+            .subviews
+            .filter{$0.accessibilityIdentifier == AccessibilityIdentifier.stackViewForImageViewAndLabels}
```
[MUST] [【TIPS】今すぐ `.filter{...}.first` をやめましょう](https://qiita.com/lovee/items/eb6ce12408d85a1bb23a)

Hyperbolic4183 committed:
fix fliter.firstを使うのをやめた

refs #33

novr reviewed (approved):
![LGTM](https://image.lgtmoon.dev/135789)
